oaknational/oak-ai-lesson-assistant
# feat: detect and upgrade users with missing metadata

## Description

Existing users were onboarded without us recording their region, which we need to determine their demo status. This PR takes these legacy users through a quick onboarding just to record the region and demo status

It's also a good opportunity to namespace the public metadata to the `labs` prefix. Existing users will already be migrated to this structure with a script

- Metadata structure
  - Move public user metadata to `publicMetadata.labs`
  - Add `isOnboarded` to public metadata
- Middleware
  - Add `isOnboarded` and `isDemoUser` to Clerk session claims to avoid server round trip
  - Detect users without `isOnboarded` session claim and redirect to /onboarding
  - Allow routes required for onboarding if the user needs to onboard
  - Dev: Allow routes with `x-dev-preload` even if they're protected
- Onboarding backend
  - Split demo user setup to new tRPC endpoint
- Onboarding frontend
  - Extract current onboarding form to a new AcceptTermsForm component
  - Check user metadata for `isOnboarded`
    - Show AcceptTermsForm if not onboarded
    - Just update demo status if already onboarded 
  - Regenerate token on completion to update claims

## How to test

1. Create or modify a test user with the following public metadata: `{ "labs": { "isOnboarded": true } }`
2. Log in with that user at https://oak-ai-lesson-assistant-k3tc8h51r.vercel-preview.thenational.academy
3. You go through a quick onboarding state before landing on the homepage
4. You have been assigned a demo status in your public metadata
5. You can see this demo status on /aila

## Screenshots

Current onboarding form. This shouls stay the same for new users 
![CleanShot 2024-08-28 at 13 40 36@2x](https://github.com/user-attachments/assets/9d5a9940-b53b-43aa-b912-85d085fc316f)

For legacy users without demo status

![CleanShot 2024-08-28 at 13 42 34@2x](https://github.com/user-attachments/assets/39659bf9-0602-47c6-8738-58bcee2bdb40)


## Checklist
